[Bug] dubbo3.3.2 produce huge warn log and with over heap memory

### Pre-check

- [X] I am sure that all the content I provide is in English.


### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dubbo/issues?q=is%3Aissue) and found no similar issues.


### Apache Dubbo Component

Java SDK (apache/dubbo)

### Dubbo Version

dubbo3.3.2,openjdk17,centos7.9

### Steps to reproduce this issue

upgrade from 3.2.11 to 3..3.2,online service throw exeption and produce  huge log and with over heap memory.
screenshot is
![1731921409662](https://github.com/user-attachments/assets/440c3295-b1c2-4197-ab9d-746f3ec11983)

and it makes my service down !!!

### What you expected to happen

need fix it quickly

### Anything else

_No response_

### Are you willing to submit a pull request to fix on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

> Please provider the logs in text format.

`Caused by: java.io.IOException: org.apache.dubbo.common.serialize.SerializationException: java.lang.IllegalArgumentException: [Serialization Security] Serialized class org.apache.dubbo.remoting.http12.exception.EncodeException is not in allow list. Current mode is `STRICT`, will disallow to deserialize it by default. Please add it into security/serialize.allowlist or follow FAQ to configure it.
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper.handleToIOException(DefaultSerializationExceptionWrapper.java:353)
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper.access$000(DefaultSerializationExceptionWrapper.java:27)
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper$ProxyObjectOutput.writeObject(DefaultSerializationExceptionWrapper.java:319)
	at org.apache.dubbo.common.serialize.DefaultMultipleSerialization.serialize(DefaultMultipleSerialization.java:34)
	at org.apache.dubbo.rpc.protocol.tri.ReflectionPackableMethod$WrapResponsePack.pack(ReflectionPackableMethod.java:330)
	at org.apache.dubbo.rpc.model.PackableMethod.packResponse(PackableMethod.java:46)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcCompositeCodec.encode(GrpcCompositeCodec.java:84)
	... 207 common frames omitted
Caused by: org.apache.dubbo.common.serialize.SerializationException: java.lang.IllegalArgumentException: [Serialization Security] Serialized class org.apache.dubbo.remoting.http12.exception.EncodeException is not in allow list. Current mode is `STRICT`, will disallow to deserialize it by default. Please add it into security/serialize.allowlist or follow FAQ to configure it.
	... 214 common frames omitted
Caused by: java.lang.IllegalArgumentException: [Serialization Security] Serialized class org.apache.dubbo.remoting.http12.exception.EncodeException is not in allow list. Current mode is `STRICT`, will disallow to deserialize it by default. Please add it into security/serialize.allowlist or follow FAQ to configure it.
	at org.apache.dubbo.common.utils.DefaultSerializeClassChecker.loadClass0(DefaultSerializeClassChecker.java:153)
	at org.apache.dubbo.common.utils.DefaultSerializeClassChecker.loadClass(DefaultSerializeClassChecker.java:109)
	at org.apache.dubbo.common.serialize.hessian2.Hessian2SerializerFactory.getDefaultSerializer(Hessian2SerializerFactory.java:56)
	at com.alibaba.com.caucho.hessian.io.SerializerFactory.loadSerializer(SerializerFactory.java:404)
	at com.alibaba.com.caucho.hessian.io.SerializerFactory.getSerializer(SerializerFactory.java:333)
	at com.alibaba.com.caucho.hessian.io.SerializerFactory.getObjectSerializer(SerializerFactory.java:284)
	at com.alibaba.com.caucho.hessian.io.Hessian2Output.writeObject(Hessian2Output.java:434)
	at org.apache.dubbo.common.serialize.hessian2.Hessian2ObjectOutput.writeObject(Hessian2ObjectOutput.java:102)
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper$ProxyObjectOutput.writeObject(DefaultSerializationExceptionWrapper.java:317)
	... 211 common frames omitted
| WARN | 2024-11-18 17:02:31 | Log4j2Logger.java | warn | 106 |  [DUBBO] Error while doOnNext, dubbo version: 3.3.2, current host: 192.168.0.220, error code: 99-0. This may be caused by , go to https://dubbo.apache.org/faq/99/0 to find instructions.  |
org.apache.dubbo.remoting.http12.exception.EncodeException: Internal Server Error
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcCompositeCodec.encode(GrpcCompositeCodec.java:90)
	at org.apache.dubbo.remoting.http12.message.HttpMessageEncoder.encode(HttpMessageEncoder.java:36)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.buildMessage(AbstractServerHttpChannelObserver.java:212)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.doOnNext(AbstractServerHttpChannelObserver.java:142)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:96)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.customizeError(AbstractServerHttpChannelObserver.java:247)
	at org.apache.dubbo.rpc.protocol.tri.h12.grpc.GrpcUnaryServerChannelObserver.customizeError(GrpcUnaryServerChannelObserver.java:30)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onError(AbstractServerHttpChannelObserver.java:110)
	at org.apache.dubbo.remoting.http12.AbstractServerHttpChannelObserver.onNext(AbstractServerHttpChannelObserver.java:99)`

--- a/dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/AbstractServerHttpChannelObserver.java
+++ b/dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/AbstractServerHttpChannelObserver.java
@@ -96,7 +96,14 @@ public abstract class AbstractServerHttpChannelObserver<H extends HttpChannel> i
             doOnNext(data);
         } catch (Throwable t) {
             LOGGER.warn(INTERNAL_ERROR, "", "", "Error while doOnNext", t);
-            onError(t);
+            Throwable throwable = t;
+            try {
+                doOnError(throwable);
+            } catch (Throwable t1) {
+                LOGGER.warn(INTERNAL_ERROR, "", "", "Error while doOnError, original error: " + throwable, t1);
+                throwable = t1;
+            }
+            onCompleted(throwable);
         }
     }
 
@@ -105,7 +112,6 @@ public abstract class AbstractServerHttpChannelObserver<H extends HttpChannel> i
         if (closed) {
             return;
         }
-
         try {
             throwable = customizeError(throwable);
             if (throwable == null) {
@@ -122,7 +128,6 @@ public abstract class AbstractServerHttpChannelObserver<H extends HttpChannel> i
             LOGGER.warn(INTERNAL_ERROR, "", "", "Error while doOnError, original error: " + throwable, t);
             throwable = t;
         }
-
         onCompleted(throwable);
     }
 

